tools/tcg_rim_tool/src/main/java/hirs/swid/CredentialParser.java

```diff
@@ -0,0 +1,190 @@
+package hirs.swid;
+
+import org.bouncycastle.cert.jcajce.JcaX509ExtensionUtils;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.openssl.PEMKeyPair;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.bouncycastle.util.encoders.Base64;
+
+import java.io.*;
+import java.security.*;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+
+/**
+ * This class parses private key, public key, and certificate for use in their respective java.security objects.
+ */
+public class CredentialParser {
+    private static final String X509 = "X.509";
+    private static final String JKS = "JKS";
+    private static final String PEM = "PEM";
+    private static final String PKCS1_HEADER = "-----BEGIN RSA PRIVATE KEY-----";
+    private static final String PKCS1_FOOTER = "-----END RSA PRIVATE KEY-----";
+    private static final String PKCS8_HEADER = "-----BEGIN PRIVATE KEY-----";
+    private static final String PKCS8_FOOTER = "-----END PRIVATE KEY-----";
+    private X509Certificate certificate;
+    private PrivateKey privateKey;
+    private PublicKey publicKey;
+
+    public X509Certificate getCertificate() {
+        return certificate;
+    }
+
+    public PrivateKey getPrivateKey() {
+        return privateKey;
+    }
+
+    public PublicKey getPublicKey() {
+        return publicKey;
+    }
+
+    public void parseJKSCredentials() {
+        KeyStore.PrivateKeyEntry privateKeyEntry =
+                parseKeystorePrivateKey(SwidTagConstants.DEFAULT_KEYSTORE_PATH,
+                        SwidTagConstants.DEFAULT_PRIVATE_KEY_ALIAS,
+                        SwidTagConstants.DEFAULT_KEYSTORE_PASSWORD);
+        certificate = (X509Certificate) privateKeyEntry.getCertificate();
+        privateKey = privateKeyEntry.getPrivateKey();
+        publicKey = certificate.getPublicKey();
+    }
+
+    public void parsePEMCredentials(String certificateFile, String privateKeyFile) throws FileNotFoundException {
+        certificate = parsePEMCertificate(certificateFile);
+
+        /*User input on algorithm???*/
+        privateKey = parsePEMPrivateKey(privateKeyFile, "RSA");
+
+        publicKey = certificate.getPublicKey();
+    }
+
+    /**
+     * This method returns the X509Certificate found in a PEM file.
+     * @param filename
+     * @return
+     * @throws FileNotFoundException
+     */
+    private X509Certificate parsePEMCertificate(String filename) throws FileNotFoundException {
+        X509Certificate certificate = null;
+        try {
+            FileInputStream fis = new FileInputStream(filename);
```

Comment:
Are these streams closed at all?

Reply:
Added a finally {...} block to close streams.